### PR TITLE
Add support for resuming training from checkpoint

### PR DESCRIPTION
**Pull request by jxl26**
## Description
Add support for resuming training from checkpoint

## Example / Current workflow
The current workflow only supports reinitializing a model with the state_dictionary from a checkpoint file, while users might want to also restore the optimizer state and gradient values at a checkpoint, so as to resume training progress from a given point.

## Bugfix / Desired workflow
The new workflow supports a --resume-from-checkpoint flag for training, which starts model training at the optimizer state stored in the checkpoint.

## Relevant issues
#1039 

## Checklist
- [x] linted with flake8?
- [x] unit tests added?

### CODE REVIEW

File: chemprop/cli/train.py [hunk 1]
```diff
@@ -1749,7 +1764,8 @@ def train_model(
                 )
 
         best_model_path = checkpointing.best_model_path
-        model = model.__class__.load_from_checkpoint(best_model_path)
+        if not args.resume_from_checkpoint:
+            model = model.__class__.load_from_checkpoint(best_model_path)
```
[the remark this answers is not in the document]
Comment (jxl26) [reply chemprop/cli/train.py]: The main issue again is the pipeline trying to fetch checkpoints produced by the previous run, which could have been on a different device and leads to error.
